bin/main.ml

```diff
@@ -71,16 +98,47 @@ let precondition_frequency_pair prop (gen_type, name) =
       func l))
 
 
+let create_test_list prop gen = 
+  List.map (precondition_frequency prop) gen
+let create_test_pair_list prop gen = List.map (precondition_frequency_pair prop) gen
+
+(* let create_safe_test_list prop gen =
+  List.map (precondition_frequency ) prop *)
 
-let create_test_list prop gen_name = List.map (precondition_frequency prop) gen_name
-let create_test_pair_list prop gen_name = List.map (precondition_frequency_pair prop) gen_name
 
-(* tests: gets run by qcheck *)
-(* maybe I should make a funtion for running tests to make it clear *)
-let tests = create_test_list is_sorted' Arbitrary_builder.sorted_list_arbitraries
+(* algebraic data type for the generator arbitraries,
+  helps with making a cleaner association list *)
+type generator_t = 
+  | Single of (int list QCheck.arbitrary * string) list
+  | Pair of ((int * int list) QCheck.arbitrary * string) list
+  | Option of (int list option QCheck.arbitrary * string) list
 
-(* foldername: where results gets outputted to if -o *)
-let foldername = "./bin/sorted_list/"
+(* association list for the generators and their names *)
+let generator_assoc_list = [ 
+  ("duplicate_list", Single Arbitrary_builder.duplicate_list_arbitraries) ;
+  ("sized_list", Pair Arbitrary_builder.sized_list_arbitraries) ;
+  ("sorted_list", Option Arbitrary_builder.sorted_list_arbitraries) ;
+  ("unique_list", Single Arbitrary_builder.unique_list_arbitraries) ;
+  ]
+  
+(* I don't know if there's something similar for functions? 
+  I found something about GADT's but I would need to investigate more *)
```

Comment:
I see how GADT's could help here, lets not do that for the sake of time. If we then need to just iterate through one list of benchmarks at a time then that is fine